From b4555aa4f2d56ff587a2e91af68477210b8e98f1 Mon Sep 17 00:00:00 2001
From: Fam Zheng <famz@redhat.com>
Date: Mon, 4 Aug 2014 09:48:45 -0400
Subject: [virt] hyperv/storvsc: Filter commands based on the storage protocol version

Message-id: <1407145729-25017-4-git-send-email-famz@redhat.com>
Patchwork-id: 86954
O-Subject: [RHEL-7 PATCH 3/7] Drivers: scsi: storvsc: Filter commands based on the storage protocol version
Bugzilla: 1122317
RH-Acked-by: Jason Wang <jasowang@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: K. Y. Srinivasan <kys@microsoft.com>

Going forward it is possible that some of the commands that are not currently
implemented will be implemented on future Windows hosts. Even if they are not
implemented, we are told the host will corrrectly handle unsupported
commands (by returning appropriate return code and sense information).
Make command filtering depend on the host version.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Reviewed-by: Hannes Reinecke <hare@suse.de>
Cc: <stable@vger.kernel.org>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 8caf92d80526f3d7cc96831ec18b384ebcaccdf0)
Signed-off-by: Fam Zheng <famz@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/storvsc_drv.c b/drivers/scsi/storvsc_drv.c
index 1fdb47f..d21c716 100644
--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -1550,9 +1550,19 @@ static int storvsc_queuecommand(struct Scsi_Host *host, struct scsi_cmnd *scmnd)
  struct vmscsi_request *vm_srb;
  struct stor_mem_pools *memp = scmnd->device->hostdata;
 
- if (!storvsc_scsi_cmd_ok(scmnd)) {
-  scmnd->scsi_done(scmnd);
-  return 0;
+ if (vmstor_current_major <= VMSTOR_WIN8_MAJOR) {
+  /*
+   * On legacy hosts filter unimplemented commands.
+   * Future hosts are expected to correctly handle
+   * unsupported commands. Furthermore, it is
+   * possible that some of the currently
+   * unsupported commands maybe supported in
+   * future versions of the host.
+   */
+  if (!storvsc_scsi_cmd_ok(scmnd)) {
+   scmnd->scsi_done(scmnd);
+   return 0;
+  }
  }
 
  request_size = sizeof(struct storvsc_cmd_request);
-- 
1.7.1